### PR TITLE
support force upgrade when pd cluster is unavailable

### LINKED ISSUE #609: fail to upgrade pd
## Bug Report

**What version of Kubernetes are you using?**

```
Client Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.2", GitCommit:"17c77c7898218073f14c8d573582e8d2313dc740", GitTreeState:"clean", BuildDate:"2018-10-24T06:54:59Z", GoVersion:"go1.10.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.2", GitCommit:"17c77c7898218073f14c8d573582e8d2313dc740", GitTreeState:"clean", BuildDate:"2018-10-24T06:43:59Z", GoVersion:"go1.10.4", Compiler:"gc", Platform:"linux/amd64"}
```
**What version of TiDB Operator are you using?**

```
TiDB Operator Version: version.Info{TiDBVersion:"2.1.0", GitVersion:"v1.0.0-beta.2.38+758c9888ea4245-dirty", GitCommit:"758c9888ea4245f4d651f8f3f95c31290131f2e4", GitTreeState:"dirty", BuildDate:"2019-06-03T11:39:25Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}
```
**What storage classes exist in the Kubernetes cluster and what are used for PD/TiKV pods?**

```
[pingcap@172.16.4.4 dashboard]$ kubectl get pvc --namespace=ryan
NAME                    STATUS   VOLUME              CAPACITY   ACCESS MODES   STORAGECLASS        AGE
pd-ryan-test-pd-0       Bound    local-pv-d381798e   3667Gi     RWO            shared-nvme-disks   12d
pd-ryan-test-pd-1       Bound    local-pv-3c881820   3667Gi     RWO            shared-nvme-disks   8d
pd-ryan-test-pd-2       Bound    local-pv-458ec73a   3667Gi     RWO            shared-nvme-disks   8d
tikv-ryan-test-tikv-0   Bound    local-pv-dc52bd99   3667Gi     RWO            nvme-disks          12d
tikv-ryan-test-tikv-1   Bound    local-pv-bd6f7b9a   3667Gi     RWO            nvme-disks          12d
tikv-ryan-test-tikv-2   Bound    local-pv-6fc44b05   3667Gi     RWO            nvme-disks          12d
tikv-ryan-test-tikv-3   Bound    local-pv-968c86d4   3667Gi     RWO            nvme-disks          12d
tikv-ryan-test-tikv-4   Bound    local-pv-6bd5a4a2   3667Gi     RWO            nvme-disks          8d
```
**What's the status of the TiDB cluster pods?**

```
[pingcap@172.16.4.4 dashboard]$ kubectl get pod -n ryan -o wide
NAME                                   READY   STATUS    RESTARTS   AGE    IP               NODE          NOMINATED NODE
ryan-test-discovery-6cd69db65f-gxpwl   1/1     Running   0          12d    10.233.73.50     172.16.4.94   <none>
ryan-test-monitor-75fc65dbd5-kfjgc     2/2     Running   0          12d    10.233.73.14     172.16.4.94   <none>
ryan-test-pd-0                         1/1     Running   0          74m    10.233.73.6      172.16.4.94   <none>
ryan-test-tidb-0                       1/1     Running   0          12d    10.233.111.206   172.16.4.87   <none>
ryan-test-tikv-0                       1/1     Running   0          6d3h   10.233.68.102    172.16.4.32   <none>
ryan-test-tikv-1                       1/1     Running   0          6d3h   10.233.73.75     172.16.4.91   <none>
ryan-test-tikv-2                       1/1     Running   0          6d3h   10.233.103.122   172.16.4.90   <none>
ryan-test-tikv-3                       1/1     Running   0          6d3h   10.233.101.76    172.16.4.35   <none>
ryan-test-tikv-4                       1/1     Running   0          6d3h   10.233.84.213    172.16.4.99   <none>
```
**What did you do?**
 

* upgrade pd to `hub.pingcap.net/pingcap/pd:rleungx-0b5e79e` 
* update pd to `hub.pingcap.net/pingcap/pd:rleungx-368ee03`  

**What did you expect to see?**
pd use this image  `hub.pingcap.net/pingcap/pd:rleungx-368ee03`  
**What did you see instead?**
```
Containers:
  pd:
    Container ID:  docker://e26a211c49a80301d3f7c669522f32a31a8bccab690ba3c98fe2b8ebd7116b68
    Image:         hub.pingcap.net/pingcap/pd:rleungx-0b5e79e
    Image ID:      docker-pullable://hub.pingcap.net/pingcap/pd@sha256:7dc1c85026ebaaa4b8206027b40b2f8e37264fffdb8e9d07e4d9c2209cc0938d
```

### PR DESCRIPTION
**Pull request by DanielZhangQD**
### What problem does this PR solve? 
close #609 
### What is changed and how does it work?
Support force upgrade if PD cluster is unavailable
### Check List 

Tests 

 - Unit test
 - E2E test
 - Stability test
 - Manual test (add detailed scripts or steps below)
    - PD cluster image pull backoff during the creation
    - PD cluster pending due to affinity setting during the creation
    - PD cluster is unavailable after creation
    - PD cluster is running normally
    - PD cluster only has 1 replica

Code changes

 - Has Go code change
 - Has documents change

Related changes

 - Need to update the documentation

### Does this PR introduce a user-facing change?:
 
```
Support force upgrade when PD cluster is unavailable.
To enable force upgrade:
kubectl annotate --overwrite tc ${releaseName} -n ${namespace} tidb.pingcap.com/force-upgrade=true
To disable force upgrade:
kubectl annotate tc ${releaseName} -n ${namespace} tidb.pingcap.com/force-upgrade-
Note: Force upgrade must be disabled when PD cluster is recovered.
```

### CODE REVIEW

File: pkg/manager/member/pd_member_manager.go [hunk 1]
```diff
@@ -200,6 +200,16 @@ func (pmm *pdMemberManager) syncPDStatefulSetForTidbCluster(tc *v1alpha1.TidbClu
 		glog.Errorf("failed to sync TidbCluster: [%s/%s]'s status, error: %v", ns, tcName, err)
 	}
 
+	if !tc.Status.PD.Synced {
+		force := needForceUpgrade(tc)
+		if force {
+			tc.Status.PD.Phase = v1alpha1.UpgradePhase
+			setUpgradePartition(newPDSet, 0)
+			errSTS := pmm.updateStatefulSet(tc, newPDSet, oldPDSet)
+			return controller.RequeueErrorf("tidbcluster: [%s/%s]'s pd needs force upgrade, %v", tc.GetNamespace(), tc.GetName(), errSTS)
```
Comment (weekface) [inline pkg/manager/member/pd_member_manager.go]:
```suggestion
			return controller.RequeueErrorf("tidbcluster: [%s/%s]'s pd needs force upgrade, %v", ns, tcName, errSTS)
```

**Comment by weekface** [comment]
This pr needs a user-facing change(release note).

File: pkg/manager/member/utils.go [hunk 2]
```diff
@@ -214,3 +216,15 @@ func CombineAnnotations(a, b map[string]string) map[string]string {
 	}
 	return a
 }
+
+// needForceUpgrade check if force upgrade is necessary
+func needForceUpgrade(tc *v1alpha1.TidbCluster) bool {
+	// Check if annotation 'pingcap.com/forceUpgrade: "true"' is set
```
Comment (weekface) [inline pkg/manager/member/utils.go]: forceUpgrade to force-upgrade

File: pkg/manager/member/pd_member_manager_test.go [hunk 3]
```diff
@@ -639,6 +639,133 @@ func TestPDMemberManagerUpgrade(t *testing.T) {
...
+		}
+	}
+	tests := []testcase{
+		{
+			name: "force upgrade",
+			modify: func(cluster *v1alpha1.TidbCluster) {
+				cluster.Spec.PD.Image = "pd-test-image:v2"
+				cluster.Spec.PD.Replicas = 1
+				cluster.ObjectMeta.Annotations = make(map[string]string)
+				cluster.ObjectMeta.Annotations["tidb.pingcap.com/forceUpgrade"] = "true"
```
Comment (weekface) [inline pkg/manager/member/pd_member_manager_test.go]: ditto

Comment (DanielZhangQD) [reply pkg/manager/member/pd_member_manager_test.go]: done

File: pkg/manager/member/utils.go (hunk 2 above)
Comment (DanielZhangQD) [reply pkg/manager/member/utils.go]: done

**Pull request merged**